### ISSUE TITLE
Cancel job has no expected effect

### ISSUE THREAD

**Issue by xmik** [gocd/gocd#3375]
### Issue Type
- Bug Report

### Summary
When there is a task being already run, I can cancel its job using GoCD UI. I expected that this would terminate the task and so the process on go-agent vm should no longer exist.

#### Reproduction
I have a small GoCD pipeline configuration in yaml:
```yaml
pipelines:
  gocd-cancel-task:
    group: gocd
    label_template: "${git[:8]}"
    materials:
      git:
        git: "git@git.example.com:gocd/gocd-cancel-task.git"
        branch: master
    stages:
      - bash_loop:
          clean_workspace: true
          jobs:
            bash_loop:
              resources:
                - docker
                - small
              tasks:
               - exec:
                  command: bash
                  arguments:
                    - -c
                    - while true; do date; sleep 1; done
```
(The git repository contains only the file with GoCD pipeline configuration).
On GoCD server it resulted in this output:
```
# removed some initial lines
11:01:35.820 [go] Start to build gocd-cancel-task/d759430b/bash_loop/1/bash_loop on small-docker-worker-1 [/var/lib/go-agent/2]
11:01:35.820 [go] Current job status: passed.

11:01:35.820 [go] Start to execute task: <exec command="bash" timeout="0" >
<arg>-c</arg>
<arg>while true; do date; sleep 1; done</arg>
</exec>.
11:01:35.824 Sat Apr  8 11:01:35 UTC 2017
11:01:36.829 Sat Apr  8 11:01:36 UTC 2017
# and so on...
11:01:57.908 Sat Apr  8 11:01:57 UTC 2017
```
Then I cancelled the job through UI and so the output is:
```
11:01:58.013 [go] Start to execute cancel task: NULL
11:01:58.013 [go] Task is cancelled
11:01:58.912 Sat Apr  8 11:01:58 UTC 2017
11:01:59.915 Sat Apr  8 11:01:59 UTC 2017
11:02:00.919 Sat Apr  8 11:02:00 UTC 2017
# and so on...
11:03:40.287 Sat Apr  8 11:03:40 UTC 2017
```
The task is still being run. I ssh'ed into the go-agent vm and there was task process still running:
```
$ ps aux | grep while
go       19446  0.0  0.0  11244  2896 ?        S    11:01   0:00 bash -c while true; do date; sleep 1; done
go       19795  0.0  0.0  12944   976 pts/0    S+   11:02   0:00 grep --color=auto while
```
It was only after I killed the process:
```
sudo kill -9 19446
```
that the task stopped printing date output and the output on GoCD server appeared then:
```
11:03:41.312 [go] Job completed gocd-cancel-task/d759430b/style_unit/1/bash_loop on small-docker-worker-1 [/var/lib/go-agent/2]
```

### Basic environment details
GoCD Server:
  * GoCD Version: 16.12.0 (1-7375246f44f19ee4f03d943e72fec553a3a7e774) (Available in the footer)
  * JVM version:	1.8.0_111 (Available from http://<go-server-host>/go/about)
  * OS Information:	Linux 3.13.0-48-generic (Available from http://<go-server-host>/go/about)

GoCD Agent
   * GoCD Version: 17.1.0-2 (from `dpkg -s go-agent`)
   * OS: `Linux small-docker-worker-1 4.4.0-45-generic #66-Ubuntu SMP Wed Oct 19 14:12:37 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux`  (from `uname -a`); Ubuntu 14.04.2

### Other
[Here](https://github.com/gocd/gocd/issues/3298#issuecomment-290699584) it's written that:
> tasks without on-cancel task will be terminated ones the stage is cancelled.

This should be applicable in this situation.

**Comment by tomzo** [comment]
This affects only tasks configured via config repo. Fix in https://github.com/gocd/gocd/pull/3463

**Comment by tomzo** [comment]
Fixed in 17.5
